### ISSUE TITLE
gen_mmu.py appears to generate incorrect tables in some situations

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#2162]
**_Reported by Andrew Boie:_**

I reproduced this issue when I was testing patches to enable the MMU in QEMU by default.
See attached patch:

```
From 32762fdcfc2fbb2f07b634a4ec7bc4958d1cd9c2 Mon Sep 17 00:00:00 2001
From: Andrew Boie <andrew.p.boie@intel.com>
Date: Wed, 28 Jun 2017 15:04:53 -0700
Subject: [PATCH] qemu: increase RAM size and enable MMU

Previously we were instantiating QEMU with 32MB of RAM but only
enabling a small fraction of it.

Now instantiate QEMU with 8MB of RAM. We ignore the first megabyte,
but will give the system 7MB of RAM in the range 0x100000 - 0x7FFFFF.

If XIP is enabled, the "ROM" region will be 0x100000 - 0x3FFFFF (3MB)
and the remaining 4MB as the "RAM" region.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>
---
 arch/x86/soc/ia32/Kconfig.defconfig    | 6 +++---
 boards/x86/qemu_x86/Makefile.board     | 2 +-
 boards/x86/qemu_x86/qemu_x86_defconfig | 1 +
 3 files changed, 5 insertions(+), 4 deletions(-)

diff --git a/arch/x86/soc/ia32/Kconfig.defconfig b/arch/x86/soc/ia32/Kconfig.defconfig
index 9231934..52a9771 100644
--- a/arch/x86/soc/ia32/Kconfig.defconfig
+++ b/arch/x86/soc/ia32/Kconfig.defconfig
@@ -18,11 +18,11 @@ config PHYS_RAM_ADDR
 	default 0x00400000
 
 config RAM_SIZE
-	default 128 if XIP
-	default 256
+	default 4096 if XIP
+	default 7168
 
 config ROM_SIZE
-	default 72 if XIP
+	default 3072 if XIP
 
 config SYS_CLOCK_HW_CYCLES_PER_SEC
 	default 150000000 if LOAPIC_TIMER
diff --git a/boards/x86/qemu_x86/Makefile.board b/boards/x86/qemu_x86/Makefile.board
index 5ceade8..db2287d 100644
--- a/boards/x86/qemu_x86/Makefile.board
+++ b/boards/x86/qemu_x86/Makefile.board
@@ -5,7 +5,7 @@ QEMU_BIOS ?= /usr/share/qemu
 QEMU_CPU_TYPE_x86 = qemu32
 QEMU_X86_NO_REBOOT_y =
 QEMU_X86_NO_REBOOT_  = -no-reboot
-QEMU_FLAGS_x86 = -m 32 -cpu $(QEMU_CPU_TYPE_x86) \
+QEMU_FLAGS_x86 = -m 8 -cpu $(QEMU_CPU_TYPE_x86) \
 		$(QEMU_X86_NO_REBOOT_$(CONFIG_REBOOT)) \
 		-nographic -vga none -display none -net none \
 		-clock dynticks -no-acpi -balloon none \
diff --git a/boards/x86/qemu_x86/qemu_x86_defconfig b/boards/x86/qemu_x86/qemu_x86_defconfig
index cfddd3b..58bd1b1 100644
--- a/boards/x86/qemu_x86/qemu_x86_defconfig
+++ b/boards/x86/qemu_x86/qemu_x86_defconfig
@@ -14,3 +14,4 @@ CONFIG_SERIAL=y
 CONFIG_UART_NS16550=y
 CONFIG_UART_CONSOLE=y
 CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC=25000000
+CONFIG_X86_MMU=y
-- 
2.9.3

```

What happens is that for tests/kernel/xip immediately after the MMU is enabled and the "jmp _Cstart" is executed, the emulator crashes to the command prompt.

Adjusting the memory values seems to make this go away. For example, if I set RAM_SIZE and ROM_SIZE to both 1024 it doesn't happen. I suspect there are some edge cases in gen_mmu.py which are causing this.

This only seems to happen with XIP configuration (where there are two memory regions within RAM).

(Imported from Jira ZEP-2328)

**Comment by nashif** [comment]
**_by Andrew Boie:_**

* [0001-qemu-increase-RAM-size-and-enable-MMU.patch](https://s3.amazonaws.com/logs.zephyrproject.org/jira/GH-2162/0001-qemu-increase-RAM-size-and-enable-MMU.patch) (size: 2239)

**Comment by nashif** [comment]
**_by Andrew Boie:_**

Note: there are other test cases that fail sanitycheck due to the above patch, but I am fixing them. This particular issue seems rooted in the page table generation.

**Comment by nashif** [comment]
**_by Andrew Boie:_**


Clone this:

https://github.com/andrewboie/zephyr/tree/gen_mmu_needs_fix

 

And try to run any test with qemu_x86, the emulator hangs or aborts back to the command line.

Looking at the debug output of gen_mmu.py (I added some more printouts to show what is in MMU_LIST), memory Region 4 looks very suspicious, we should not need a region starting at 0x00800000 since RAM ends at 0x007FFFFF.

Need a fix for this ASAP.

 

5 memory ranges found in MMU_LIST:

Range                 Size       Flags

0xfed00000-0xfed00fff 0x00001000 0x00000002

0xfec00000-0xfecfffff 0x00100000 0x00000002

0xfee00000-0xfee00fff 0x00001000 0x00000002

0x00004000-0x007fffff 0x007fc000 0x00000002

0x00001000-0x00003fff 0x00003000 0x00000000

 

Total Page directory entries 4

Memory Region 1 start address = 0x4000

Memory Region 2 start address = 0x1000

Memory Region 3 start address = 0x400000

Memory Region 4 start address = 0x800000

Memory Region 5 start address = 0xfed00000

Memory Region 6 start address = 0xfec00000

Memory Region 7 start address = 0xfee00000

**Comment by nashif** [comment]
**_by Andrew Boie:_**

fixed by PR #750
